src/tensorwaves/optimizer/callbacks.py

```diff
@@ -0,0 +1,158 @@
+"""Collection of loggers that can be inserted into an optimizer as callback."""
+
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import IO, Iterable, List, Optional
+
+import tensorflow as tf
+import yaml
+from tqdm import tqdm
+
+from tensorwaves.interfaces import Estimator
+
+
+class Callback(ABC):
+    @abstractmethod
+    def __call__(self, parameters: dict, estimator_value: float) -> None:
+        pass
+
+    @abstractmethod
+    def finalize(self) -> None:
+        pass
+
+
+class ProgressBar:
+    def __init__(self) -> None:
+        self.__progress_bar = tqdm()
+
+    def __call__(self, parameters: dict, estimator_value: float) -> None:
+        self.__progress_bar.set_postfix({"estimator": estimator_value})
+        self.__progress_bar.update()
+
+    def finalize(self) -> None:
+        self.__progress_bar.close()
+
+
+class YAMLSummary:
+    def __init__(
+        self,
+        filename: str,
+        estimator: Estimator,
+        step_size: int = 10,
+    ) -> None:
+        """Log fit parameters and the estimator value to a `tf.summary`.
+
+        The logs can be viewed with `TensorBoard
+        <https://www.tensorflow.org/tensorboard>`_ via:
+
+        .. code-block:: bash
+
+            tensorboard --logdir logs
+        """
+        self.__iteration = 0
+        self.__step_size = step_size
+        self.__stream = open(filename, "w")
+        if not isinstance(estimator, Estimator):
+            raise TypeError(f"Requires an in {Estimator.__name__} instance")
+        self.__estimator_type: str = estimator.__class__.__name__
+
+    def __call__(self, parameters: dict, estimator_value: float) -> None:
+        self.__iteration += 1
+        if self.__iteration % self.__step_size != 0:
+            return
+        output_dict = {
+            "Time": datetime.now(),
+            "Iteration": self.__iteration,
+            "Estimator": {
+                "Type": self.__estimator_type,
+                "Value": float(estimator_value),
+            },
+            "Parameters": {
+                name: float(value) for name, value in parameters.items()
+            },
+        }
+        _empty_file(self.__stream)
+        yaml.dump(
+            output_dict,
+            self.__stream,
+            sort_keys=False,
+            Dumper=_IncreasedIndent,
+            default_flow_style=False,
+        )
```

Comment:
As mentioned in #154, I think it would be more useful to have a history of the fit result instead of just writing the latest state. But there is probably some preference of mine in it. Anyways, if we leave it like this (saving only the latest result ), then it could be useful to mention that in the docstring

Reply:
See also #166

Reply:
Sounds good, but the current two summaries are covering most of the needs for now already, so its not that urgent. I'm not requesting this change. Just wanted to point out. Though we should mention that this yaml summary is a bit different than the standard loggers, which use appending.

Reply:
#159 is read for review